src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs

```diff
@@ -48,5 +52,41 @@ private Matcher CreateMatcher(IReadOnlyList<Endpoint> endpoints)
 
             return builder.Build();
         }
+
+        // Used to tie the lifetime of a DataSourceDependentCache to the service provider
+        public class Lifetime : IDisposable
+        {
+            private readonly object _lock = new object();
+            private DataSourceDependentCache<Matcher> _cache;
+            private bool _disposed;
+
+            public DataSourceDependentCache<Matcher> Cache
+            {
+                get => _cache;
+                set
+                {
+                    lock (_lock)
+                    {
+                        if (_disposed)
+                        {
+                            value?.Dispose();
+                        }
+
+                        _cache = value;
+                    }
+                }
+            }
+
+            public void Dispose()
+            {
+                lock (_lock)
+                {
+                    _cache?.Dispose();
+                    _cache = null;
```

Comment:
I don't think so because when you set `Cache` it would immediately dispose the value. This is really about the possibility that someone will dispose the services while the data source matcher is being initialized. There isn't any code that sets this value over and over.

Reply:
ok!